### ISSUE TITLE
Enable codesigning by default in the simulator

### ISSUE THREAD

**Issue by rolfbjarne** [dotnet/macios#18469]
It seems the simulator won't load launch screens unless the app is signed.

Ref: https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1834468

Since most apps have launch screens, it might make sense to enable code signing in the simulator by default.

Note: we should look into what Xcode does if you don't have a Developer Account.

However, some additional testing might be required, because simply setting `EnableCodeSigning=true` doesn't work as expected:

```shell
$ dotnet build /p:EnableCodeSigning=true
[...] error : The "Codesign" task was not given a value for the required parameter "SigningKey", nor was there a "CodesignSigningKey" metadata on the resource [...]
```

OK, let's try giving the build a code signing key:

```shell
$ dotnet build /p:EnableCodeSigning=true /p:CodesignKey='Apple Development: Rolf Kvinge (CZ9YXM2X9A)'
[...] error : The "Codesign" task was not given a value for the required parameter "SigningKey", nor was there a "CodesignSigningKey" metadata on the resource [...]
```

No change!

what gets us passed that error is actually setting CodesignRequireProvisioningProfile=true, although it runs into another one:

```shell
$ dotnet build /p:EnableCodeSigning=true /p:CodesignRequireProvisioningProfile=true
[...]
_CodesignVerify:
  /usr/bin/codesign --verify -vvvv "-R=anchor apple generic and certificate 1[field.1.2.840.113635.100.6.2.1] exists and (certificate leaf[field.1.2.840.113635.100.6.1.2] exists or certificate leaf[field.1.2.840.113635.100.6.1.4] exists)" bin/iPhoneSimulator/Debug/TestLaunchScreen.iOS.app
  bin/iPhoneSimulator/Debug/TestLaunchScreen.iOS.app: valid on disk
  bin/iPhoneSimulator/Debug/TestLaunchScreen.iOS.app: satisfies its Designated Requirement
  test-requirement: code failed to satisfy specified code requirement(s)
error MSB6006: "codesign" exited with code 3.
```

So what actually makes the app build successfully is passing DisableCodesignVerification=true as well:

```shell
$ dotnet build /p:EnableCodeSigning=true /p:CodesignRequireProvisioningProfile=true /p:DisableCodesignVerification=true
[...]
```

From looking at the source code, this is what happens (don't take this as truth, it needs verification when implementing any fixes) for the simulator in the DetectSigningIdentity task:

* If a provisioning profile is needed, then the task will report '-' as the signing identity. 	 
    * Bug A: _Even if a specific signing identity was specified_. I'm not sure if's ever a valid thing to pass a specific signing identity on the simulator though, testing would be needed in XCode
* If a provisioning profile is not needed, then the task will not return anything as the signing identity.
    * Bug B: _Even if a specific signing identity was specified_. This is rather odd behavior no matter how you look at it - if a signing identity is always wrong for the simulator, then at least a warning should be reported, otherwise the signing identity should be used.

Setting CodesignRequireProvisioningProfile=true got us passed these two issues, but there's a third:

* Bug C: the signing process does not produce a verifiable app bundle when using '-' as the signing identity, so we should probably skip the CodesignVerify task in that case (or adjust what's verified).

# TLDR;

Add the following to the csproj to work around this:

```xml
<PropertyGroup>
    <EnableCodeSigning>true</EnableCodeSigning>
    <CodesignRequireProvisioningProfile>true</CodesignRequireProvisioningProfile>
    <DisableCodesignVerification>true</DisableCodesignVerification>
</PropertyGroup>
```

This should work on device as well:
* `EnableCodeSigning` defaults to true for device, so nothing changes.
* `CodesignRequireProvisioningProfile` also defaults to true for device
* `DisableCodesignVerification`: assuming the code signing configuration in the project is correct, it shouldn't matter that the verification is skipped (and if it's broken, the app will fail to install on device anyways).

**Comment by ithiredguns** [comment]
Didn't work for Xamarin Forms adding the above in property Group .. Also, I signed the App with a Distribution Certificate and with an AdHoc Profile and distributed it through AppCenter ... It still has the black splash screen 
![downloadFromAppCenter](https://github.com/xamarin/xamarin-macios/assets/27841167/cc947ed6-28fe-4617-91e8-0432bd4c173f)

https://github.com/xamarin/xamarin-macios/assets/27841167/2008dadc-9503-4bae-86e6-acd55ab7d7b4

**Comment by rolfbjarne** [comment]
@ithiredguns if the problem occurs on device, then it's a different problem (because device builds are always signed, there's no way to install unsigned builds on device), so please open a new issue and attach a test project we can use to reproduce the problem.

**Comment by ithiredguns** [comment]
The issue is still on simulators also... I did add the changes to the ios project file ...

**Comment by ithiredguns** [comment]
Wow ! Thanks for the feedback @rolfbjarne .  What worked was changing the pdf images to png, shrinking the images and also, **moving** the images away from Assets.xcassets to Resources folder and updating the LaunchScreen.storyboard to reflect the new png files (instead of pdf) , delete the app & restarting the simulator, cleaning the iOS folder by deleting the bin & obj ... Looks like Apple restricts the size of LaunchScreen images to be less than 25 MB ...

**Comment by FreakyAli** [comment]
When I do this in Maui it forces me to use a developer certificate even when I just want to run on a simulator... Which I don't have right now...

**Comment by rolfbjarne** [comment]
> When I do this in Maui it forces me to use a developer certificate even when I just want to run on a simulator... Which I don't have right now...

AFAIK there's no way around this right now if you want the launch screen to show up. When we fix this issue, we'll investigate and see if there's a way to implement code signing without requiring a developer account.

**Comment by FreakyAli** [comment]
> > When I do this in Maui it forces me to use a developer certificate even when I just want to run on a simulator... Which I don't have right now...
> 
> AFAIK there's no way around this right now if you want the launch screen to show up. When we fix this issue, we'll investigate and see if there's a way to implement code signing without requiring a developer account.

Damn that sucks, anyways can you please update here once there is a solution for this or even a workaround i guess

**Comment by rolfbjarne** [comment]
> * Bug C: the signing process does not produce a verifiable app bundle when using '-' as the signing identity, so we should probably skip the CodesignVerify task in that case (or adjust what's verified).

If I try to validate the signature of an app built for the simulator by Xcode it doesn't pass our code signing verification either, so it seems ignoring the verification step in this case is the right thing to do.

**Comment by rolfbjarne** [comment]
I've experimented a bit with Xcode, and found the following:

* I have not been able to use a specific signing certificate, the generic `-` certificate is always used.
* I have not been able to specify any custom entitlements, the one and only entitlement is com.apple.security.get-task-allow=true no matter how many capabilities I add to the Xcode project. This would also mean no provisioning profile is used either.

**Comment by last-Programmer** [comment]
I tried the workaround mentioned in https://github.com/xamarin/xamarin-macios/issues/19430#issuecomment-1814479679 for the simulator and it does not work for me. Even when deployed to ios device the launch screen is not showing. This is happening for me in .net native ios 8 project in visual studio windows. I have the developer certificate and profile set for the both simulator and device and it does not work for me.

**Comment by zijianhuang** [comment]
I have tested on real iPad of browserstack, 11, 12 and 13 are OK with: 
```xml
	<ItemGroup>
		<MauiSplashScreen Include="Resources\storyboard1536x2048.svg" />
	</ItemGroup>
```

  , but 14 and above give me black screen .

on real iPhone, the same.

Therefore, it is a problem in 14, 15 and 16, while the latest xcode supports only minimum 14.

**Comment by ToolmakerSteve** [comment]
FYI: When I write/run a simple native (Swift or Objective C) app in Xcode, SplashScreen (LaunchScreen) is shown on Simulator.

I don't know whether this ability means Xcode does do codesigning for Simulator in this case, or whether there is some other detail that Xcode knows to do for recent iOS versions, but the build script sent from Visual Studio lacks.

I mention this because before I saw the issue about codesigning, I was thinking that there was merely something new that VS needs to send to Xcode, for this to work.

**Comment by ToolmakerSteve** [comment]
FYI: Despite years of developing in .Net for iOS (at first via Xamarin Native / iOS), I had trouble recently getting Splash Screen to work. Posted a StackOverflow Q & A with the steps that I went through:

[[Using .Net 8 for iOS [NOT MAUI], add a LaunchScreen with logo image and background color](https://stackoverflow.com/questions/79565533/using-net-8-for-ios-not-maui-add-a-launchscreen-with-logo-image-and-backgrou)]
.

**Comment by FreakyAli** [comment]
@ToolmakerSteve  Hey, you got this working somehow?

**Comment by ToolmakerSteve** [comment]
> [@ToolmakerSteve](https://github.com/ToolmakerSteve) Hey, you got this working somehow?

Yes. Added answer: https://stackoverflow.com/a/79565661/199364

BOTTOM LINE: Rofl's lines to "fake" code-sign work, to show on simulator.

To get the storyboard file correct, I used Xcode to create a tiny native app (Swift app, specifying "storyboard" for LaunchScreen). Used Xcode's UI to modify LaunchScreen.storyboard. (There's an auto-constraint button somewhere at bottom, that sets the constraints to fill the screen minus "safe" margins.) Copied LaunchScreen.storyboard from that, into my project's `Resources` folder. Along with the `.png` referenced by that storyboard.

Troubleshooting: Sometimes simulator fails to show updated launch screen. (Including not showing the change from "black screen" to "now I have the needed files - works on device".) I list various steps that might be needed. Launch screen is tricky, because device or simulator seems to cache for each app that has been run. Power off/on helps.

CAVEAT: the "fake" code-signing (for Debug build) may not work on a physical device. Next step is Conditions to handle Simulator vs actual device, use different code-sign lines.

CAVEAT # 2: I Seem to need certificate on both mac and pc. (Might be related to how this Apple team was originally set up.) Still trying to get a good "recipe" for the certificate+provisioning process. However, once it works for a device attached via USB to the Mac (Building on PC via Pair to Mac), then getting it to work on simulator now seems to be easy, by changing to those "fake" code-sign lines.
